Harmonize watchers endpoints

NobodysNightmare opened the pull request:
:warning: Please merge this into `release/4.3` :warning: 
## OpenProject work package

https://community.openproject.org/work_packages/17755
## Description

This PR harmonizes the documentation around the watchers endpoint and other similar endpoints.
It also applies the relevant changes to the implementation and aligns the implementation with the specification.

NobodysNightmare committed:
specify endpoint for current watchers

[ci skip]

NobodysNightmare committed:
require permissions to see watchers

[ci skip]

NobodysNightmare committed:
clarify 403/404 for WP scoped resources

[ci skip]

NobodysNightmare committed:
change the way watchers are added

- using links
- possibility for multiple watchers at a time
- always returns all watchers (pagination stilll possible)

Note that the term "link object" is introduced by an as of now pending
PR

[ci skip]

NobodysNightmare committed:
specify further error cases

[ci skip]

NobodysNightmare committed:
fix example link

[ci skip]

NobodysNightmare committed:
bring available watchers in line with assignees etc

- endpoint also scoped by project
- thus also returning all possible watchers, not only the ones "not
watching"

[ci skip]

NobodysNightmare committed:
clarification on 403/404 of available xxx

[ci skip]

NobodysNightmare committed:
also return 422 for non-watchable users

[ci skip]

NobodysNightmare committed:
Merge remote-tracking branch 'origin/dev' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

[ci skip]

NobodysNightmare committed:
rename data to payload, reference it in "links"

[ci skip]

NobodysNightmare committed:
only allow to add one watcher at a time

NobodysNightmare committed:
specify error id for type mismatch of links

[ci skip]

NobodysNightmare committed:
fixed example

[ci skip]

NobodysNightmare committed:
Merge branch 'release/4.2' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

NobodysNightmare committed:
return ResourceTypeMismatch for wrong resources

- this is according to new specification changes
- I have no clue why the initial author did this "wrap into a validation error" business

NobodysNightmare committed:
adapt watch/unwatch links to specification

- they are called (un)watch, not (un)watchChanges
- watch link indicates the required payload using the payload attribute
- required payload already indicated using new (as of this commit invalid) format

NobodysNightmare committed:
remove redundant authentication

already performed on project level

NobodysNightmare committed:
style resource link parser

NobodysNightmare committed:
update roar version

NobodysNightmare committed:
extract parsable link into own representer

NobodysNightmare committed:
accept new request format for add watcher

also adapt the angular frontend to speak the new format

NobodysNightmare committed:
avoid useless JSON roundtrip

know your library they said...

NobodysNightmare committed:
fix specs to respect new error identifier

NobodysNightmare committed:
clarify more edge cases regarding watchers

NobodysNightmare committed:
try to improve readability of watcher spec

incl. Rubocop fixes

NobodysNightmare committed:
allow to view list of current watchers

NobodysNightmare committed:
fix HTTP 500 in case of not allowed watcher

NobodysNightmare committed:
also use named parameters for RemoveWatcher service

NobodysNightmare committed:
specify and test edge cases for watcher deletion

reconsidered when to return 404:

while the watcher resource does technically not exist when a user is not watching,
it seems more client friendly to still return a success code even if nothing was done

Still I differentiated that from user that does not exist at all

NobodysNightmare committed:
further specification refinements

- remove build warnings
- move available watchers endpoint back to WP scope

NobodysNightmare committed:
correctly link & embed watchers in WP

NobodysNightmare committed:
add spec for new link object

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	config/locales/de.yml
	config/locales/en.yml
	spec/lib/api/v3/work_packages/work_package_representer_spec.rb

ulferts commented:
This got me quite confused so please ask if my writing is not understandable:

It used to be my understanding that the `watch` link on work package was designed to add oneself as the watcher of the work package. Apparently, that understanding was wrong - there seems to have been the possibility to add arbitrary users at this endpoint. At least that what it says in the docu. The code says differently. But the new code at least allows for adding arbitrary watchers via the endpoint. 

What do we need the `addWatcher` link for then? Is it to signal, that the user can add arbitrary users instead of only oneself? The addWatcher link is described wrongly as it has the href `"/api/v3/work_packages/1528/watchers{?user_id}"` and states to be templated. The code however does not support providing the user_id as a parameter. Only payload is allowed. 

The permissions checked for the `watch` link read as if that link is only shown if it should be used for adding oneself. The `add_work_package_watchers` permission is not checked there. It is checked for the `addWatcher` link but that link does not take into account whether the user has the `view_work_packages` permission which should suffice for adding oneself.

The `addWatcher` link could probably be removed if it weren't for the need to signal to the front end whether the user is allowed to add watchers other than oneself. This could of course be mitigated by the front-end fetching the list of available_watchers and noting that he is the only one on that list but it would be awkward and error prone (there might be only the one user in the project for now).

On top of this confusion, and the links being displayed when they should not, there is a bug in combination with the front end. When I click on the star icon in the work package split view I can no longer watch the work package. That again is due to the user_id parameter.

NobodysNightmare committed:
simplify conditions + specs

if a user is not allowed to view a work package he should not get to see
the WP representer. (even hypothetically: then watcher links would not be the prime problem)

Therefore the conditions can be simplified as well as the corresponding specs

NobodysNightmare committed:
remove hardcoded titles

titles are specified as optional, but when they are present they should be localized

NobodysNightmare committed:
also provide a payload for addWatcher

allow payloads to be templated (to some degree) as well

NobodysNightmare committed:
fix toggle watch code

floriank commented:
this could either use a rebase and some love or a close.

NobodysNightmare commented:
It needs a merge and love, **not** a close ^^

I will do that _soon_.

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb
	lib/api/v3/work_packages/work_package_representer.rb
	lib/api/v3/work_packages/work_packages_api.rb

Azure7111 commented:
The specs seem to fail legitimately or at least somewhat consistently.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -114,12 +115,17 @@ For example a work package without an assignee will still have an assignee link,
 
+If a `payload` is specified, it needs to be sent as the body of the request to achieve the desired result (e.g. perform the
+action represented by the link). If no `payload` is specified, there is either no payload to send or there are multiple valid
+payloads. A payload might also be templated slightly. If the `templated` property is true, a payload might contain replaceable parts
```
"...templated _slightly_"? I'd say either something is templated or it isn't.

NobodysNightmare replied:
It is not fully templatable, I only added semantics for templating inside strings in the payload.

If you know a better word, tell me...

Azure7111 replied:
Hmm, I don't care that much, but how about "parameters"?

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3172,21 +3180,21 @@ the human readable name of custom fields.*
+                            "user": { "href": "/api/v3/users/1" }
+                        }
                     },
                     "addWatcher": {
```
What about "removeWatcher"? Since the ticket does not explicitly state the necessity of either one I'd expected it for the sake of consistency. Also: Should both actions go to the same endpoint but use different http verbs (like POST and DELETE) or rather use separate endpoints?

NobodysNightmare replied:
sure :cat2:

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3430,306 @@ For more details and all possible responses see the general specification of [Fo
++ Response 403 (application/hal+json)
 
-## Add watcher [POST]
+    Returned if the work package does not exist or the client does not have sufficient permissions to see it.
```
Copy pasta? :spaghetti:  Here we should already have the permission to see a work package but lack the permission to edit it.

NobodysNightmare replied:
mmmhhhhh, pasta....

:spaghetti: :spaghetti: :spaghetti: :spaghetti:

NobodysNightmare committed:
fix c&p error

[ci skip]

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3430,306 @@ For more details and all possible responses see the general specification of [Fo
+                                "title": "Set lock on j.sheppard"
+                                "method": "POST"
+                            },
+                            "delete": {
```
I know this is only a slight thing and probably just my own insecurity speaking. I also think this would take more work than might be appropriate, but: As I read this link my first thought was: "And this is probably a link to remove this user as watcher from this WP" because I am looking at a collection of watchers. Immerdiately thereafter I realized that this is wrong, e.g. because it makes no sense to "lock" a watcher. But: What do you think about postfixing the link names with "User"? (i.e. "lockUser" and "deleteUser")

I only wrote about this point because my misunderstanding might have made me delete a user, whereas I only wanted to remove a watcher.

NobodysNightmare replied:
I see your point.
I would argue to not address it now, because those link names make perfect sense on a user object alone and are also already in use.

What might make sense in a future revision of the API: Returning watcher objects instead of user objects (e.g. `api/v3/watchers/13`). Database side watchers have a unique id (e.g. if you watch two work packages, there are two watcher objects for your user account) and might make it more clear _what_ you are deleting.

However, the initial thought of not exposing watcher objects was to provide an abstraction from the database concept and call everything a user. This is now one caveat.

**tl;dr:** For now I would leave it as-is, but if it turns out that there will be more confusion that gains, we might want to reverse course...

Azure7111 replied:
Okay as we agreed: We will not do anything about this until the amount of inadvertently deleted users becomes too damn high.

NobodysNightmare committed:
add a removeWatcher link to work packages

NobodysNightmare committed:
fix spec

the actual hashes provided by grape have string keys

NobodysNightmare committed:
clarify specification

NobodysNightmare committed:
style

[ci skip]

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
+The request is expected to contain a single JSON object, that contains a link object under the `user` key.
+
+The response will be user added as watcher.
+In case the user was already watching the work package an `HTTP 200` is returned, an
```
Have you considered to place this remark into the section of the corresponding response?

NobodysNightmare replied:
I remember something like the blueprint not working when I do that...

edit: [found it](https://github.com/opf/openproject/commit/89ef5f648fe264ef82358405b69dd030d28ce1c4)

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
+
+    Returned if the client does not have sufficient permissions.
+
+    **Required permission:**
```
It's more than one ;)

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
+                "message": "Expected resource of type 'User', but got a 'Status'.",
+            }
+
 ## Remove Watcher [/api/v3/work_packages/{work_package_id}/watchers/{id}]
```
I know you did not touch this but what about renaming id -> user_id?

NobodysNightmare replied:
I guess there will be more occurences of such errors.

One (anyone) could do this in a ticketless cleanup PR IMO, I am too lazy to do it in this PR TBH.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
+
+    **Required permission:** view work package
+
+    Or the specified user does not exist at all.
```
That would be a third case, you might to consider rephrasing your sentence above which states that there are only two cases. Also: Maybe pull this sentence, or its statement for that matter, above the line with Required permission.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
+                "message": "The specified work package does not exist."
+            }
+
+## Available Watchers [/api/v3/work_packages/{id}/available_watchers]
```
It appears to me as if this is a more general inconsistency in our API documentation: Somewhere above you went the extra mile and called this `id` actually `work_package_id`.

NobodysNightmare replied:
It is usually the follwoing rule:
- `id` if it is onyl one
- `specific_id` for the extra thing needed, in case it is more than one

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3492,10 +3803,10 @@ For more details and all possible responses see the general specification of [Fo
 
 + Parameters
-    + work_package_id (required, integer, `1`) ... Work package id
+    + id (required, integer, `1`) ... work package id
```
Could you enlighten me with the rationale for this change?

NobodysNightmare replied:
As stated above:

Use the shortest possible name and only prepend words if that is neccessary.

If one wanted to change that, I would propose a separate PR that changes it consistently across the board.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3492,10 +3803,10 @@ For more details and all possible responses see the general specification of [Fo
 ## Available watchers [GET]
 
-Gets a list of users that can watch the work package.
+Gets a list of users that are able to be watchers of the specified work package.
```
This is totally fine with me, but what is the gain here?

NobodysNightmare replied:
No idea...

edit: probably history of changes... or any other thought that crossed my mind...

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3507,6 +3818,8 @@ Gets a list of users that can watch the work package.
 
     **Required permission:** add work package watchers
 
+    *Note that you will only receive this error, if you are at least allowed to see the corresponding project.*
```
I bet this is :spaghetti: Copy pasta again. This should probably read "work package" instead of "project".

NobodysNightmare replied:
No copy pasta, but change pasta.
This endpoint was moved from project to work package and the message was not updated.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -3515,6 +3828,20 @@ Gets a list of users that can watch the work package.
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "The specified project does not exist."
```
This seems weird, I'd have expected this entire response to revolve around the specified _work package_.

NobodysNightmare replied:
the same copy pasta you noticed above... not sure why this comment did not get outdated...

Azure7111 commented on lib/api/decorators/link_object.rb:
```diff
@@ -0,0 +1,70 @@
+
+                 return nil unless id
+
+                 api_v3_paths.send(@path, id)
```
Would this work for paths which do not require an id?

NobodysNightmare replied:
no

NobodysNightmare committed:
fix doc error caused by moving the endpoint

endpoint was moved from project scope to WP scope

[ci skip]

NobodysNightmare committed:
Merge remote-tracking branch 'upstream/release/4.2' into fix/specify_watchers_link

Conflicts:
	spec/requests/api/v3/watcher_resource_spec.rb

Azure7111 commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -301,13 +309,18 @@ def activities
+          self_link = api_v3_paths.work_package_watchers(represented.id)
+
+          # FIXME/LEGACY: we pass the WP as context?!? that makes a difference!!!
+          # tl;dr: the embedded user representer must not be better than any other user representer
```
That sounds indeed odd. But have you found the reason why?

NobodysNightmare replied:
The user representer actually makes use of the pased work package and provides more links in that case (watcher specific links).

This is in contrast to everything we discussed today about separating between user and watcher. However, I found it out of scope of this (already grown) PR to also address this issue, because at least our angular client relies on those fields ATM.

ulferts commented:
@NobodysNightmare we should no longer merge this into 4.2. So once this is supposed to be done, please reissue against release/4.3

Azure7111 commented on lib/api/v3/work_packages/work_packages_shared_helpers.rb:
```diff
@@ -38,26 +38,21 @@ def request_body
-              merge_json_into_work_package!(work_package, request_body.to_json)
-            rescue ::API::Errors::Form::InvalidResourceLink => e
-              fail ::API::Errors::Validation.new(e.message)
-            end
```
Could you shed some light on this removal?

NobodysNightmare replied:
Sure:

There was a special case for one possible exception that could be thrown from the merge-method. Just to wrap this specific exception in another exception.

However, the cleaner approach is to immediately use the original exception here, as (per specification of this PR) there should be a separate error identifier for the given class of errors.

NobodysNightmare replied:
Thinking about it now: This might be a bad thing because now that error will not cause a `Validation` error. Which in case is checked by [some other code](https://github.com/opf/openproject/blob/dev/lib/api/v3/work_packages/work_packages_shared_helpers.rb#L89).

I guess I will come up with a solution for that in my other PR: https://github.com/opf/openproject/pull/3166

Azure7111 commented on spec/lib/api/v3/work_packages/work_package_representer_spec.rb:
```diff
@@ -493,7 +483,15 @@
+        it 'should have a link to remove watcher' do
+          expect(subject).to be_json_eql(
+                               api_v3_paths.watcher('{user_id}', work_package.id).to_json)
+                               .at_path('_links/removeWatcher/href')
```
This indentation looks weird 0o. I'd expected these lines to be either more to the right, which might not be possible due to the line limit, or to the left (indented only once).

But since @houndci didn't have to say anything about this I'm a little confused and this is not important at all.

Azure7111 commented on spec/lib/api/v3/work_packages/work_package_representer_spec.rb:
```diff
@@ -507,6 +505,39 @@
+          role.permissions.delete(:delete_work_package_watchers) and role.save
+        end
+
+        it 'should not have a link to add watcher' do
```
The test seems fine, its name not so much ;)

Azure7111 commented on spec/lib/api/v3/work_packages/work_package_representer_spec.rb:
```diff
@@ -507,6 +505,39 @@
+
+      describe 'watchers link' do
+        context 'when the user is allowed to see watchers' do
+          it_behaves_like 'has an untitled link' do
```
I know that that the title of a link is optional, but is there a specific reason why we want to ensure that this one has no title?

NobodysNightmare replied:
Specific tests are good.
I did not implement a title so I expect none.

If someone now decides to add a shitty title he will at least have to add a shitty spec too...

Azure7111 commented on spec/lib/api/v3/work_packages/work_package_representer_spec.rb:
```diff
@@ -507,6 +505,39 @@
+          end
+
+          it 'embeds the watchers as collection' do
+            is_expected.to be_json_eql('Collection'.to_json).at_path('_embedded/watchers/_type')
```
At this point I could imagine having a shared example checking the integrity of a collection. I'm not sure if we have something like that already. What do you think about that?

NobodysNightmare replied:
We have that already, however I don't want to test the representer for user collection here. I only want to verify that the wp representer utilizes such a representer and then there is another test checking whether the user collection representer does sane things.

Now I only need an approximation for "the wp representer uses a user collection representer in apropriate places". The approximation I've chosen here (and I think in other places too) is that there should be the type `Collection` at a specific path.

An evil developer could make this test green by re-implementing the user collection representer inline in the wp representer. But evil developers are hard to detect via unit tests ;-)

Azure7111 commented on spec/lib/api/v3/work_packages/work_packages_shared_helpers_spec.rb:
```diff
@@ -67,19 +67,23 @@ def status(_code)
 
         it 'does not raise for validation errors' do
-          expect(subject.validation_errors.any?).to be_truthy
+          expect(subject.validation_errors).not_to be_empty
```
Nice catch! :+1:

NobodysNightmare committed:
clarify words

[ci skip]

Azure7111 commented on spec/requests/api/v3/watcher_resource_spec.rb:
```diff
@@ -182,47 +226,21 @@
-      watcher_ids = watcher_hashes.map { |watcher_hash| watcher_hash['id'] }
-
-      expect(watcher_ids).to match_array([current_user.id, available_watcher.id])
+    it 'includes a user eligable for watching' do
```
typo

NobodysNightmare committed:
style

[ci skip]

NobodysNightmare commented:
CI is as green as it will get ^^

NobodysNightmare commented on spec/lib/api/decorators/link_object_spec.rb:
```diff
@@ -0,0 +1,128 @@
+
+        it 'throws an error' do
+          expect { representer.from_hash parsed_hash }.to raise_error(
+                                                            ::API::Errors::InvalidResourceLink)
```
> 1 file inspected, no offenses detected

@houndci is a liar!

NobodysNightmare committed:
typo

[ci skip]

Azure7111 commented:
This PR has been merged into release/4.3, as seen in b3a8e0187ec4aad998dc66d669abb52531e66166.
